### ISSUE TITLE
question 

### ISSUE THREAD

**Issue by hacker-wei** [ahangchen/TFusion#12]
![default](https://user-images.githubusercontent.com/32689594/45064770-300f2400-b0e8-11e8-8c14-13fd810ce551.png)
i DON'T know this directory how to get .could you help me ?

**Comment by ahangchen** [comment]
- GRID as Target Dataset

  - follow dataset instruction, split the dataset to ten cross-validation sets
in each cross-validation set, rename training directory to 'train', rename probe directory to 'probe', 
  - renmae gallery directory to 'test'
  - you can also refer to 'TrackViz/data/grid' for more details about GRID cross validation.

**Comment by hacker-wei** [comment]
![default](https://user-images.githubusercontent.com/32689594/45072966-d3276400-b10f-11e8-88c1-5019f24a9020.png)
![default](https://user-images.githubusercontent.com/32689594/45072984-e5a19d80-b10f-11e8-94d6-efa06c97f017.png)
first picture is your code, rank_reid/dataset don't have duke_train.txt ,but have grid_train.txt .what should i do?

**Comment by ahangchen** [comment]
sources contains the names of source dataset that you want to transfer from. They are configurable. If you only want to reproduce the experiments in CVPR version. You can just remove 'duke'.

**Comment by hacker-wei** [comment]
ok.thank you.

**Comment by hacker-wei** [comment]
![default](https://user-images.githubusercontent.com/32689594/45154918-cbe58080-b20b-11e8-840a-9e9f22bff1ed.png)
I can't found this sorce_train.list and source dataset, and pair_transfer.py

**Comment by ahangchen** [comment]
these *.h5 files are generated by pair_train.py, pair_transfer.py is not used for CVPR version. You can follow ctrl/transfer.py pipeline to understand how these files work together.

**Comment by hacker-wei** [comment]
![default](https://user-images.githubusercontent.com/32689594/45160966-54b8e800-b21d-11e8-9a2c-e75f2b5fbcf7.png)

I know *.h5 are generated by pair_train.py,but I couldn't found source_train.list

**Comment by ahangchen** [comment]
They are listed in [rank-reid/dataset](https://github.com/ahangchen/rank-reid/tree/dev/dataset) dir. Please follow ctrl/transfer.py for more details.

**Comment by hacker-wei** [comment]
![default](https://user-images.githubusercontent.com/32689594/45260350-0b80b800-b418-11e8-8165-884555850587.png)
请问作者，为什么我这个文件夹下面的什么都没有，导致我后面出现错误，后面几个里面都有，只是-r-train文件夹里面没有文件，好奇怪

**Comment by hacker-wei** [comment]
你好，作者 。不知道您什么时候发布您的稳定版的代码？我的market-grid 和 viper-grid分别是35 和 50，差距有点大。不知道该如何调节参数。

**Comment by ahangchen** [comment]
viper2grid能到50的话market2grid也能到比较高才对啊，能贴一下两组迁移对应论文table1的几个结果吗？

**Comment by hacker-wei** [comment]
[grid_eval.log](https://github.com/ahangchen/TFusion/files/2374998/grid_eval.log)

**Comment by hacker-wei** [comment]
[market_result_eval.log](https://github.com/ahangchen/TFusion/files/2375001/market_result_eval.log)

**Comment by ahangchen** [comment]
-  target为market的结果跟论文里的基本一致呀，融合和增量都有提升并且达到了论文里的AP
- target为grid的结果有几个有点怪
  - viper_grid：cv7增量前融合后(line 60)，迷之下降，但其他还是符合之前做的实验，也基本达到了论文的AP
  - market_grid: cv1-增量后融合后(line 96)，迷之下降，但其他结果还是基本符合预期的，目测了一下十个划分平均结果肯定不止35啊
 - cuhk2grid：基本符合预期

其实TFusion里的版本应该是当时最后一个版本的，也是比较稳定的，但是经常有同学直接去下另外两个仓库的master分支就会遇到许多奇怪的问题，你这里复现的结果基本都是对的，可能是上边两处操作上有问题，代码参数应该不用改，可以重新跑一下对应的那两个划分看看。

**Comment by hacker-wei** [comment]
你的意思就是，我单独跑viper_grid：cv7和market_grid: cv1是不是？

**Comment by ahangchen** [comment]
是的，TrackViz/ctrl/img_st_fusion.py可以作为这种单划分实验的入口，要改一下main，觉得麻烦的话，也可以直接跑ctrl/transfer.py，source只保留viper和Market，target只保留grid

**Comment by hacker-wei** [comment]
[grid_eval.log](https://github.com/ahangchen/TFusion/files/2378013/grid_eval.log)
作者，您看，这是我第二次单独跑了之后的结果，有没有问题？

**Comment by ahangchen** [comment]
还是那两个划分的结果不太对，不过十个划分平均后影响不大，也可以就在这个基础上做研究

> 估算随机时空模型有两种做法，一种是选取所有样本对算时间差，一种是随机选取train_cnt*useful_predict_limit个样本对算时间差，当useful_predict_limit数量等于训练样本数时，与前一种等价，为了减少计算量，目前是让useful_predict_limit等于max(train_cnt/100, 10)，算是对随机时空分布的一个近似，所以会导致grid这种小数据集上结果有波动，但是grid有十个划分，所以影响不大

**Comment by hacker-wei** [comment]
知道了，谢谢你

**Comment by hacker-wei** [comment]
![image](https://user-images.githubusercontent.com/32689594/45595282-a874bf80-b9db-11e8-8a8f-eec3b224210d.png)
请问这个地方为什么要判断ig_cnt>=YES[0],这是表示什么意思？
![image](https://user-images.githubusercontent.com/32689594/45595300-f38ed280-b9db-11e8-9313-376abf049105.png)
这个地方也很懵逼，不知道啥意思？cnt和hit表示什么？求作者解答，谢谢

**Comment by ahangchen** [comment]
这个是从 market1501的MATLAB评估代码改过来的，也参考了 hehefan的 [python版本](https://github.com/hehefan/Unsupervised-Person-Re-identification-Clustering-and-Fine-tuning/blob/master/baseline/evaluate.py#L95)， hit就是预测的结果命中正样本的数目，cnt是命中前已预测的样本数目（可以用来算precision），其中同摄像头同ID的样本对和ID为-1的样本对，不算正样本也不算负样本，因此有ig_cnt这种变量，统计需要忽略的样本数目。

**Comment by hacker-wei** [comment]
> They are listed in [rank-reid/dataset](https://github.com/ahangchen/rank-reid/tree/dev/dataset) dir. Please follow ctrl/transfer.py for more details.

**Comment by hacker-wei** [comment]
谢谢

**Comment by hacker-wei** [comment]
作者，您好，我还想问下，这里的result_argsort是不是一开始的target dataset的测试排序的结果？而且这里是不是保存的是二维矩阵，行表示query，列表示test里面查找的下标排序？

**Comment by ahangchen** [comment]
- 这里的result_argsort是不是一开始的target dataset的测试排序的结果？
- 是
- 而且这里是不是保存的是二维矩阵，行表示query，列表示test里面查找的下标排序？
- 是

**Comment by hacker-wei** [comment]
好的，谢谢，代码确实有点难度啊，确实很厉害，佩服

**Comment by hacker-wei** [comment]
![image](https://user-images.githubusercontent.com/32689594/45726316-3d9cd180-bbf1-11e8-898d-13203ecc9dfc.png)
作者，还得麻烦您一下，这里确实有点模糊

**Comment by ahangchen** [comment]
```python
for ig in IGNORE:
    # YES[0]表示第1个正确样本的位置，ig_cnt统计的是在第一次命中之前有多少个可忽略样本
    if ig < YES[0]:
        ig_cnt += 1
    else:
        break
if ig_cnt >= YES[0]:  # 实际上由于数据特点，这个分支只有等于会成立，大于号是为了让下面那个else逻辑上显式只包括小于的情况
    # ig_cnt = YES[0]意味着第一次命中之前预测的样本都是可忽略的，所以可以认为rank1命中了，同理可以修改YES[0]为YES[5]的计算rank5
    rank_1 += 1
    rank1_list.append(1)
else:
    # 在第一次正确命中之前有其他不可忽略的错误预测，说明rank1没有命中
    rank1_list.append(0)

for i, k in enumerate(YES):
    ig_cnt = 0
    for ig in IGNORE:
        if ig < k:
            ig_cnt += 1
        else:
            break
    cnt = k + 1 - ig_cnt  # k+1表示已预测的个数，由于有需要忽略的样本，所以减去ig_cnt
    hit = i + 1  # 命中个数
    tmp_recall = hit / len(YES)
    tmp_precision = hit / cnt
    ap = ap + (tmp_recall - recall) * ((precision + tmp_precision) / 2)  # 梯形逼近PR曲线面积
    recall = tmp_recall
    precision = tmp_precision
```

**Comment by hacker-wei** [comment]
作者你好，我想看看在你的代码上修改一下提取特征的网络，换成其他的网络，例如：InceptionV3网络，但是Inceptionv3网络的最后dropout前的输出的维度是（None,5,5,2048）,resnet50是（None，1，1,2048）：
（1）我在训练某个数据集的时候，生成的softmax.h5和pair.h5文件都没问题，后面报错是维度的问题，在这里我做了修改，在模型的输出部分加了，池化的操作，将其转化为（1,1,1,2048），对于模型外和模型内进行池化，这没区别吧？可是不知道为什么我那个里面始终没有训练得到的score.txt 和 pid.txt?导致我test找不到这两个文件而报错。希望作者能给我建议
![image](https://user-images.githubusercontent.com/32689594/45799894-a9f6fe00-bce1-11e8-8ec4-01396e9a96ea.png)
![image](https://user-images.githubusercontent.com/32689594/45799996-01956980-bce2-11e8-9d67-97ba9c4925f9.png)

**Comment by ahangchen** [comment]
光看这部分代码好像是没问题的，得找具体的写文件那块代码调试看看

**Comment by hacker-wei** [comment]
作者你好，我稍微修改了一个您的代码，就是在pair_train那一部分，加了之后我训练后始终收敛不了，不知道为啥原因，不知道您能不能够帮我看下代码可行性

**Comment by superCman** [comment]
我跟你有一样的问题，我将作者pair_train的欧式距离函数改变之后，jude的损失一直持续在0.5无法下降，请问你解决了问题吗？

**Comment by hacker-wei** [comment]
@superCman 我改的不是欧式距离这块，你看你改了之后有没有将距离归一化到0-1范围之内

**Comment by superCman** [comment]
请问它那个欧氏距离只有平方，没有开根号是不是有错呀?

**Comment by superCman** [comment]
请问你方便留一个联系方式吗，想请教一下你问题。

**Comment by superCman** [comment]
@hacker-wei

**Comment by hacker-wei** [comment]
@superCman 可以，你留方式给我，我加你

**Comment by superCman** [comment]
我的qq是247321025，麻烦你加一下。 @hacker-wei

**Comment by geloveyou** [comment]
@hacker-wei 请问可以添加一下联系方式吗？有问题想请教您

**Comment by geloveyou** [comment]
@hacker-wei  我的qq 是1462220177

**Comment by XB-70** [comment]
> @superCman我改的不是欧式距离这块，你看你改了之后有没有将距离归一化到0-1范围之内

请问可以加一下联系方式吗？复现遇见了一些问题，想请教一下

**Comment by XB-70** [comment]
> ![默认](https://user-images.githubusercontent.com/32689594/45064770-300f2400-b0e8-11e8-8c14-13fd810ce551.png)
> 我不知道该目录如何获取。您能帮我吗？

请问可以加一下联系方式吗？复现遇见了一些问题，想请教一下，我的QQ是1526377816
